src/Libraries/Microsoft.AspNetCore.Telemetry.Middleware/Logging/Internal/HttpLoggingRedactionInterceptor.cs

```diff
@@ -0,0 +1,206 @@
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Telemetry.Internal;
+using Microsoft.Extensions.Telemetry.Logging;
+
+namespace Microsoft.AspNetCore.Telemetry.Http.Logging;
+
+internal sealed class HttpLoggingRedactionInterceptor : IHttpLoggingInterceptor
+{
+    private const string RequestStartTimestamp = "RequestStartTimestamp";
+    private const string Durration = "duration";
+
+    // These three fields are "internal" solely for testing purposes:
+    internal TimeProvider TimeProvider = TimeProvider.System;
+
+    private readonly IncomingPathLoggingMode _requestPathLogMode;
+    private readonly HttpRouteParameterRedactionMode _parameterRedactionMode;
+    private readonly ILogger<HttpLoggingRedactionInterceptor> _logger;
+    private readonly IHttpRouteParser _httpRouteParser;
+    private readonly IHttpRouteFormatter _httpRouteFormatter;
+    private readonly IIncomingHttpRouteUtility _httpRouteUtility;
+    private readonly HeaderReader _requestHeadersReader;
+    private readonly HeaderReader _responseHeadersReader;
+    private readonly string[] _excludePathStartsWith;
+    private readonly IHttpLogEnricher[] _enrichers;
+    private readonly FrozenDictionary<string, DataClassification> _parametersToRedactMap;
+
+    public HttpLoggingRedactionInterceptor(
+        IOptions<LoggingRedactionOptions> options,
+        ILogger<HttpLoggingRedactionInterceptor> logger,
+        IEnumerable<IHttpLogEnricher> httpLogEnrichers,
+        IHttpRouteParser httpRouteParser,
+        IHttpRouteFormatter httpRouteFormatter,
+        IRedactorProvider redactorProvider,
+        IIncomingHttpRouteUtility httpRouteUtility)
+    {
+        var optionsValue = options.Value;
+        _logger = logger;
+        _httpRouteParser = httpRouteParser;
+        _httpRouteFormatter = httpRouteFormatter;
+        _httpRouteUtility = httpRouteUtility;
+
+        _parametersToRedactMap = optionsValue.RouteParameterDataClasses.ToFrozenDictionary(StringComparer.Ordinal, optimizeForReading: true);
+
+        _requestPathLogMode = EnsureRequestPathLoggingModeIsValid(optionsValue.RequestPathLoggingMode);
+        _parameterRedactionMode = optionsValue.RequestPathParameterRedactionMode;
+
+        _requestHeadersReader = new(optionsValue.RequestHeadersDataClasses, redactorProvider);
+        _responseHeadersReader = new(optionsValue.ResponseHeadersDataClasses, redactorProvider);
+
+        _excludePathStartsWith = optionsValue.ExcludePathStartsWith.ToArray();
+
+        _enrichers = httpLogEnrichers.ToArray();
+    }
+
+    public void OnRequest(HttpLoggingContext logContext)
+    {
+        var context = logContext.HttpContext;
+        var request = context.Request;
+        if (ShouldExcludePath(context.Request.Path))
+        {
+            logContext.LoggingFields = HttpLoggingFields.None;
+        }
+        else if (logContext.IsAnyEnabled(HttpLoggingFields.Response))
```

Comment:
Turning bits on is fine, so long as an interceptor is internally consistent. E.g. this interceptor only stores the start time if it expects a response. It may need a little hardening in case the key isn't found later for any reason. I don't know that I expect many competing interceptors.

Reply:
> I don't know that I expect many competing interceptors.

I don't expect many. Mostly I think we are reliant people only using one, or taking some care when selecting/ordering multiple to avoid conflicts. If it ever does scale up further it can be a future problem.

> It may need a little hardening in case the key isn't found later for any reason

Yeah I would, or unconditionally capture the data.